cmd/terramate/cli/cli.go

```diff
@@ -597,14 +604,59 @@ func (c *cli) cloneStack() {
 }
 
 func (c *cli) generate(workdir string) {
-	report := generate.Do(c.cfg(), workdir)
+	report, vendorReport := c.gencodeWithVendor(workdir)
+
 	c.output.Msg(out.V, report.Full())
 
+	vendorReport.RemoveIgnoredByKind(download.ErrAlreadyVendored)
+
+	if !vendorReport.IsEmpty() {
+		c.output.Msg(out.V, vendorReport.String())
+	}
+
+	// TODO(KATCIPIS): what if vendorReport has errors/ignored ? exit 1 ?
 	if report.HasFailures() {
 		os.Exit(1)
 	}
 }
 
+// gencodeWithVendor will generate code for the whole project providing automatic
+// vendoring of all tm_vendor calls.
+func (c *cli) gencodeWithVendor(workdir string) (generate.Report, download.Report) {
+	vendorProgressEvents := download.NewEventStream()
+	progressHandlerDone := c.handleVendorProgressEvents(vendorProgressEvents)
+
+	vendorRequestEvents := make(chan event.VendorRequest)
+	vendorReports := download.HandleVendorRequests(
+		c.prj.root,
+		vendorRequestEvents,
+		vendorProgressEvents,
+	)
+
+	mergedVendorReport := download.MergeVendorReports(vendorReports)
+
+	log.Debug().Msg("generating code")
+
+	report := generate.Do(c.cfg(), workdir, c.vendorDir(), vendorRequestEvents)
+
+	log.Debug().Msg("stopping vendor request handler")
```

Comment:
Yeah, this one is for debug situation where odd things are happening, like maybe something si getting stuck forever (which already happened :laughing: ), the message does say it is "stopping", so if something still happens after the log I don't think it is misleading, stopping is always about sending the stop message (like closing the channel), and then we wait for confirmation it stopped (in that case I think then I log something "now it stopped" -P). But no strong felings on any of this, feel free to propose different/better messages :100:

Reply:
Hmm I get it, on my side, thinking on this:

"Having said that, if I see the message stopping vendor request handler then either we are abruptly stopping the "vendor request handler" or the "vendor request handler" itself finished and is now stopping"

I'm thinking more in terms of process A is sending a message for process B to stop, which is the channel being closed, it doesn't seem as confusing to me because in Go there is no way to abruptly stop another goroutine anyway, you can just send a message instructing it to stop, like sending a signal, etc. If the message said the goroutine is stopped then it would be misleading for sure, because it isn't (like you said), but Im just indicating that the stopping process started.

When the actual goroutine stops then we can have a debug saying it stopped. So I would see on the logs:

* "Stopping X"
* Some things happened
* "X stopped/finished"

So in general I don't find it misleading, but I also like the other message, will change it then :+1: